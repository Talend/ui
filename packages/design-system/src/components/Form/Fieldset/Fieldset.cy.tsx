/* eslint-disable testing-library/await-async-query */
/* eslint-disable testing-library/prefer-screen-queries */
import { useState } from 'react';
import Fieldset from './Fieldset';
import { ButtonPrimary } from '../../Button';

context('<Fieldset />', () => {
	it('should preserve children component state between renders', () => {
		const TestComponentWithState = () => {
			const [value] = useState(Math.random());
			return <div data-testid="random-value-on-mount">{value}</div>;
		};

		const Wrapper = () => {
			const [hasError, setHasError] = useState(false);
			return (
				<>
					<Fieldset>
						{hasError ? <span>Error message</span> : null}
						<TestComponentWithState />
					</Fieldset>
					<ButtonPrimary onClick={() => setHasError(!hasError)}>Toggle error</ButtonPrimary>
				</>
			);
		};

		cy.mount(<Wrapper />);
		cy.findByTestId('random-value-on-mount')
			.invoke('val')
			.then(value => {
				cy.get('button').click();
				cy.findByTestId('random-value-on-mount').should('contain', value);
			});
	});
});
