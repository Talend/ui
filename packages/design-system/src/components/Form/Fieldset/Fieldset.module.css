.fieldset {
  padding: 0;
  border: none;
}

.legend {
  color: var(--coral-color-neutral-text);
  padding: var(--coral-spacing-m) 0;
  font: var(--coral-heading-m);
  margin: 0;
}

.fieldset-content {
  display: flex;
  flex-flow: column;
  gap: var(--coral-spacing-s);
}