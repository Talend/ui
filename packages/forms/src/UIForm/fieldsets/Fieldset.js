import React, { PropTypes } from 'react';
import Widget from '../Widget';

export default function Fieldset(props) {
	const { schema, ...restProps } = props;
	const { title, items } = schema;

	return (
		<fieldset className="form-group">
			{title && (<legend>{title}</legend>)}
			{items.map((itemSchema, index) => (
				<Widget
					{...restProps}
					key={index}
					schema={itemSchema}
				/>
			))}
		</fieldset>
	);
}

Fieldset.propTypes = {
	schema: PropTypes.shape({
		items: PropTypes.array.isRequired,
		title: PropTypes.string,
	}).isRequired,
};
