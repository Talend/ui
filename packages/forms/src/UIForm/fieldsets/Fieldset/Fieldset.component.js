import PropTypes from 'prop-types';
import React from 'react';
import classnames from 'classnames';
import Widget from '../../Widget';

export default function Fieldset(props) {
	const { schema, ...restProps } = props;
	const { title, items, options } = schema;

	return (
		<fieldset className="form-group">
			<legend className={classnames({ 'sr-only': options && options.hideTitle })}>{title}</legend>
			{items.map((itemSchema, index) => (
				<Widget {...restProps} key={index} schema={itemSchema} />
			))}
		</fieldset>
	);
}

if (process.env.NODE_ENV !== 'production') {
	Fieldset.propTypes = {
		schema: PropTypes.shape({
			title: PropTypes.string.isRequired,
			items: PropTypes.array.isRequired,
		}).isRequired,
	};
}
