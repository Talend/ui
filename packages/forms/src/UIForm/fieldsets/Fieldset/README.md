# Fieldset

This widget allows you to group your fields into fieldsets.

**Json Schema**
```json
{
  "type": "object",
  "title": "Comment",
  "properties": {
    "lastname": {
      "type": "string"
    },
    "firstname": {
      "type": "string"
    },
    "age": {
      "type": "number"
    },
    "comment": {
    "type": "string"
    }
  }
}
```

**UI Schema**

| Property | Description |
|---|---|
| widget | Value : `fieldset` |
| title | The title to display above fieldset |
| items | Array of widget definitions |

```json
[
  {
    "widget": "fieldset",
    "title": "My awesome USER form",
    "items": [
      {
        "key": "lastname",
        "title": "Last Name"
      },
      {
        "key": "firstname",
        "title": "First Name"
      },
      {
        "key": "age",
        "title": "Age"
      }
    ]
  },
  {
    "widget": "fieldset",
    "title": "Rest of the form",
    "items": [
      {
        "key": "comment",
        "widget": "textarea",
        "title": "Comment"
      }
    ]
  }
]
```

**Result**

![Fieldsets](screenshot.png)
