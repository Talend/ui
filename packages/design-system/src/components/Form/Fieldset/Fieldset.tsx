import React from 'react';
import { isElement } from 'react-is';
import * as S from './Fieldset.style';

export type FieldsetProps = React.FieldsetHTMLAttributes<HTMLFieldSetElement> & {
	legend?: string;
	required?: boolean;
	readOnly?: boolean;
};

const Fieldset = React.forwardRef(
	(
		{ legend, children, disabled, readOnly, required, ...rest }: FieldsetProps,
		ref: React.Ref<HTMLFieldSetElement>,
	) => {
		const childrenProps = readOnly ? { readOnly: true } : {};
		return (
			<S.Fieldset className="c-fieldset" disabled={disabled} {...rest} ref={ref}>
				{legend && (
					<S.Legend className="c-fieldset__legend">
						{legend}
						{required && '*'}
					</S.Legend>
				)}
				{React.Children.toArray(children).map(child =>
					isElement(child) ? React.cloneElement(child, childrenProps) : child,
				)}
			</S.Fieldset>
		);
	},
);

export default Fieldset;
