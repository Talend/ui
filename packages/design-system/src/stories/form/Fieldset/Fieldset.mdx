import { Canvas, Meta, Story } from '@storybook/blocks';
import { FigmaImage } from '@talend/storybook-docs';
import { Status } from '../../Status.block';

import * as Stories from './Fieldset.stories';

<Meta of={Stories} />
<Status id="formFieldset" />

# Form Fieldset

Fieldsets group together fields within a form.

It's a semantic group, not necessarily a visual one. This component however provides a `legend` prop to help visually segment fieldsets.

## Zoning

<FigmaImage
	src="https://www.figma.com/file/qc8oUSyVJM67gSsbTShftt/%F0%9F%93%90-Zoning?node-id=349%3A227"
	alt="zoning image for form fieldset"
/>

## Style

Group logical blocks into a fieldset.

If the fields are even more closely related, display them on the same line using `Row`.

<FigmaImage
	src="https://www.figma.com/file/FaYIWpJeP6LwjWr78L8eAr/Forms?node-id=191%3A352"
	alt="Fieldset mockup"
/>

<Canvas>
	<Story of={Stories.Default} />
</Canvas>

## States

### Error

A fieldset usually does not carry error state. Its individual fields do.

<FigmaImage
	src="https://www.figma.com/file/FaYIWpJeP6LwjWr78L8eAr/Forms?node-id=224%3A13927"
	alt="Fieldset mockup with state"
/>

<Canvas>
	<Story of={Stories.Error} />
</Canvas>

In case of more than one error or a horizontal fieldset, display the errors under each affected field.

<FigmaImage
	src="https://www.figma.com/file/FaYIWpJeP6LwjWr78L8eAr/Forms?node-id=227%3A0"
	alt="Fieldset mockup with errors in two fields"
/>

<Canvas>
	<Story of={Stories.Errors} />
</Canvas>

## Content

### Legend

Provide a legend to the fieldset to help contextualize its contents.

### Inline message

Field validation may sometime be more complex than on a single field.

If the broken validation rule benefits from a single point of explanation, use an `InlineMessage` within the affected `Fieldset` instead.

This is the only exception to the rule. Errors must otherwise be displayed under each field.

<FigmaImage
	src="https://www.figma.com/file/FaYIWpJeP6LwjWr78L8eAr/Forms?node-id=141%3A11352"
	alt="inline message position for form fieldset"
/>

## Interactions

### Conditional fieldsets

A conditional fieldset appears only when previous conditions are met.

The trigger can be any "on / off" interactions available in a form:

- A checkbox has been activated
- A radio button has been selected
- A switch has been toggled
- A select field has been set to a specific value
- A button has been pressed

Use sparingly as it makes a form harder to predict and understand at a glance.

<Canvas>
	<Story of={Stories.ConditionalFieldset} />
</Canvas>

## Usage

N/A

## Accessibility

N/A
