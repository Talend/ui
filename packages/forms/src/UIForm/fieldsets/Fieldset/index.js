import Fieldset from './Fieldset.component';
import FieldsetTextMode from './displayMode/TextMode.component';

export { FieldsetTextMode };
export default Fieldset;
