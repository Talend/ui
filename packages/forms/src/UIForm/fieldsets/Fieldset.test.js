import React from 'react';
import { shallow } from 'enzyme';

import Fieldset from './Fieldset';

describe('Fieldset widget', () => {
	it('should render fieldset', () => {
		// given
		const schema = {
			title: 'My fieldset',
			items: [
				{
					key: ['user', 'firstname'],
					type: 'text',
					schema: { type: 'string' },
				},
				{
					key: ['user', 'lastname'],
					type: 'text',
					schema: { type: 'string' },
				},
			],
		};

		// when
		const wrapper = shallow(<Fieldset schema={schema} />);

		// then
		expect(wrapper.node).toMatchSnapshot();
	});
});
