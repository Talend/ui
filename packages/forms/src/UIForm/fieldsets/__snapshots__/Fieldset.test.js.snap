exports[`Fieldset widget should render fieldset 1`] = `
<fieldset
  className="form-group">
  <legend>
    My fieldset
  </legend>
  <Widget
    schema={
      Object {
        "key": Array [
          "user",
          "firstname",
        ],
        "schema": Object {
          "type": "string",
        },
        "type": "text",
      }
    }
    widgets={Array []} />
  <Widget
    schema={
      Object {
        "key": Array [
          "user",
          "lastname",
        ],
        "schema": Object {
          "type": "string",
        },
        "type": "text",
      }
    }
    widgets={Array []} />
</fieldset>
`;
